File: src/com/company/java_examples/RotateImage.java
package com.company.java_examples;

import java.util.Arrays;

public class RotateImage {

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}
        };

        rotate(matrix);

        for (int[] row : matrix) {
            System.out.println(Arrays.toString(row));
        }
    }

    public static void rotate(int[][] matrix) {
        transpose(matrix);
        flipHorizontally(matrix);
    }

    private static void transpose(int[][] matrix) {
        for (int row = 0; row < matrix.length; row++) {
            for (int column = row; column < matrix[row].length; column++) {
                swap(matrix, row, column, column, row);
            }
        }
    }

    private static void flipHorizontally(int[][] matrix) {
        for (int row = 0; row < matrix.length; row++) {
            for (int column = 0; column < matrix[row].length / 2; column++) {
                swap(matrix, row, column, row, matrix[row].length - 1 - column);
            }
        }
    }

    private static void swap(int[][] matrix, int row1, int column1, int row2, int column2) {
        int temp = matrix[row1][column1];
        matrix[row1][column1] = matrix[row2][column2];
        matrix[row2][column2] = temp;
    }
}

File: src/com/company/kotlin_examples/RotateImage.kt
package com.company.kotlin_examples

import java.util.*

object RotateImage {

    @JvmStatic
    fun main(args: Array<String>) {
        val matrix = arrayOf(arrayOf(1, 2, 3), arrayOf(4, 5, 6), arrayOf(7, 8, 9))
        rotate(matrix)
        for (row in matrix) {
            println(Arrays.toString(row))
        }
    }

    fun rotate(matrix: Array<Array<Int>>) {
        transpose(matrix)
        flipHorizontally(matrix)
    }

    private fun transpose(matrix: Array<Array<Int>>) {
        for (row in matrix.indices) {
            for (column in row until matrix[row].size) {
                swap(matrix, row, column, column, row)
            }
        }
    }

    private fun flipHorizontally(matrix: Array<Array<Int>>) {
        for (row in matrix.indices) {
            for (column in 0 until matrix[row].size / 2) {
                swap(matrix, row, column, row, matrix[row].size - 1 - column)
            }
        }
    }

    private fun swap(matrix: Array<Array<Int>>, row1: Int, column1: Int, row2: Int, column2: Int) {
        val temp = matrix[row1][column1]
        matrix[row1][column1] = matrix[row2][column2]
        matrix[row2][column2] = temp
    }
}